### ISSUE TITLE
Getting file descriptor count on macOS throws

### ISSUE THREAD

**Issue by Therzok** [dotnet/runtime#65254]
### Description

A known way to get the file descriptor count on macOS is by using `/dev/fd`. Here is an example implementation in [Rust](https://github.com/larksuite/perf-monitor-rs/blob/299d66284603e74be5d699c8bc816574e5d581fa/src/fd/macos.rs#L4-L7).

The problem is that some file descriptors listed under `/dev/fd` are not `stat`-able, so they will fail when queried. One example is `kqueue`s.

### Reproduction Steps

`Directory.EnumerateFileSystemEntries("/dev/fd").Count();` on macOS. 

### Expected behavior

The number of open file descriptors.

### Actual behavior

```
C# > Directory.EnumerateFileSystemEntries("/dev/fd").Count();
╭─❌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ System.UnauthorizedAccessException: Access to the path '/dev/fd/5' is denied.                                                          │
│  ---> System.IO.IOException: Bad file descriptor                                                                                       │
│    --- End of inner exception stack trace ---                                                                                          │
│    at System.IO.FileStatus.ThrowOnCacheInitializationError(ReadOnlySpan`1 path)                                                        │
│    at System.IO.Enumeration.FileSystemEntry.get_IsSymbolicLink()                                                                       │
│    at System.IO.Enumeration.FileSystemEntry.Initialize(FileSystemEntry& entry, DirectoryEntry directoryEntry, ReadOnlySpan`1           │
│ directory, ReadOnlySpan`1 rootDirectory, ReadOnlySpan`1 originalRootDirectory, Span`1 pathBuffer)                                      │
│    at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()                                                                          │
│    at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)                                                                      │
│    at Submission#3.<<Initialize>>d__0.MoveNext()                                                                                       │
│ --- End of stack trace from previous location ---                                                                                      │
│    at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2   │
│ currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)                        │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

### Regression?

Not sure if it's a regression.

### Known Workarounds

Manually PInvoke `opendir` and then iterate with `readdir`. 

### Configuration

dotnet 6.0.101, macOS 10.14+, x64 and arm64

**Comment by Therzok** [comment]
`area-System.IO`, I guess :D

**Comment by danmoseley** [comment]
@Therzok what do you propose here? Presumably in the general case, we do want to throw this exception.

**Comment by Therzok** [comment]
The API implicitly does the evaluation of dir entries and their metadata. Something like directory entry count shouldn't populate the stat information of files, unless it's explicitly requested.

The equivalent C code is something like:
```c
DIR *dir = opendir("/dev/fd");
struct dirent *entries;

int count = 0;
while ((entries = readdir(dir)) {
    count++;
}
```

I'd understand if something like `GetFileSystemEntries` would do implicit evaluation here to have a snapshot of the filesystem data, but `EnumerateFileSystemEntries` is advertised as lazy. Shouldn't the entry properties be populated lazily?

**Comment by adamsitnik** [comment]
This code has been recently refactored by @tmds:

What we have in 6.0:

https://github.com/dotnet/runtime/blob/3d345e01239b31e16c426aa022763d81417d6cdb/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEntry.Unix.cs#L45-L61

What we have in 7.0:

https://github.com/dotnet/runtime/blob/e23382c32bf03d3fea5a2f1157a99148a51ca3f4/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEntry.Unix.cs#L43-L58

in theory it should not throw, let me start my Linux VM and check it

**Comment by tmds** [comment]
> This code has been recently refactored by @tmds:

Yes, this was refactored in https://github.com/dotnet/runtime/pull/60214.
One thing that was missing is `continueOnError: true` for `IsSymbolicLink` which is the cause for this exception.

**Comment by adamsitnik** [comment]
I've added a test that ensures that the bug is not coming back: https://github.com/dotnet/runtime/pull/65301

**Comment by kasperk81** [comment]
> One thing that was missing is `continueOnError: true` for `IsSymbolicLink` which is the cause for this exception.

is this fix planned for backporting to net6.0? if not, any workaround for users?

**Comment by tmds** [comment]
> if not, any workaround for users?

Not using .NET APIs.

Maybe you can try PInvoking the `proc_pidinfo` API mentioned in this article: https://zameermanji.com/blog/2021/8/1/counting-open-file-descriptors-on-macos/.

**Comment by adamsitnik** [comment]
I was not able to reproduce it using `net5.0`, it's a regression that has been introduced in `net6.0` and it's specific to macOS.

@tmds how difficult would it be to backport part of #60214 (the bug fix without refactor) to 6.0?

**Comment by tmds** [comment]
> it's a regression that has been introduced in net6.0

It was introduced in https://github.com/dotnet/runtime/pull/52235

> and it's specific to macOS.

Yes, and also specific to the special directory of `/dev/fd` which represents the open file descriptors of the process.
One of these file descriptors doesn't support `stat` probably because its something special that is not ordinarily found in a real file system.

> how difficult would it be to backport part of https://github.com/dotnet/runtime/pull/60214 (the bug fix without refactor) to 6.0?

https://github.com/dotnet/runtime/pull/60214 is a bugfix PR, it doesn't do much refactoring. There are little things to be left out. I think you should take it as a whole.

There was a separate PR to refactor some things: https://github.com/dotnet/runtime/pull/62721.

**Comment by adamsitnik** [comment]
@tmds please excuse me, for some reason I mixed these two PRs

**Comment by adamsitnik** [comment]
> is this fix planned for backporting to net6.0? if not, any workaround for users?

@kasperk81 in what circumstances have you hit this bug? Is it exactly the same use case?

**Comment by adamsitnik** [comment]
I have not received any answers for a month, so I am going to close the issue (it's fixed in 7.0, we currently don't plan to backport it to 6.0)

**Comment by Therzok** [comment]
None of the questions were directed towards me, so I'm pretty sure the issue should still be open, especially given it's a regression in net6.0

**Comment by Jozkee** [comment]
> it's a regression that has been introduced in net6.0 and it's specific to macOS.
@danmoseley @jeffhandley does this meets the bar for backporting to 6.0? 
I will check later if I can repro this on macOS.

**Comment by jeffhandley** [comment]
@Therzok Could you describe the type of application this is impacting for you and whether you'd be able to adopt a preview release of .NET 7.0 to gain the fix? Additionally, it would be helpful if you could validate that the preview .NET 7 releases do in fact address the regression in your scenario. These data help us assess the risk/reward of backporting a fix.

Thanks!

**Comment by jeffhandley** [comment]
Thank you, @Therzok. Since you have the (albeit unfortunate) workaround in place are are unblocked, and the backport isn't 100% straightforward/contained, I'm not inclined to backport the fix. If others report the same issue in a way that they are blocked, that would encourage the backport though.
